src/main/java/org/jenkinsci/bytecode/Transformer.java

```diff
@@ -57,16 +58,28 @@ public synchronized void loadRules(Collection<? extends ClassLoader> loaders) th
             return image;
 
         final ClassReader cr = new ClassReader(image);
-        final ClassWriter cw = new ClassWriter(/*ClassWriter.COMPUTE_FRAMES|*/ClassWriter.COMPUTE_MAXS);
+        final ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES|ClassWriter.COMPUTE_MAXS);
 
         final boolean[] modified = new boolean[1];
 
-        cr.accept(new ClassVisitor(ASM5,cw) {
+        // If code contains JSR/RET instructions then ASM fails to transform it with
+        // java.lang.RuntimeException: JSR/RET are not supported with computeFrames option
+        // so inline any JSR subroutines
```

Comment:
This is actually what the [AMS guys suggest](http://mail-archive.ow2.org/asm/2008-11/msg00007.html).

It *should* produce identical behavior - if you take out the code that you would have jumped to in JSR and the RET and inline it I see no reason the behaviour would be different.  That said - it did make verification of the bytecode harder and some security flaws made it though by the use of JSR/RET so in these cases it may be that code that was previously passed verification no longer does (depends exactly how much ASM rewrites and how much it keeps.